### ISSUE TITLE
parsing expressions: Maximum recursion depth exeeded or segmentation faults

### ISSUE THREAD

**Issue by wohnjayne** [sympy/sympy#10951]
I am using sympy for symbolic regression. Expressions are generated as strings, and parsed by sympy.
As soon as I introduce more complex operators such as sin(), sympy fails to parse expressions (both sympy 1.0 and 0.7.4.1).

e.g., the string
`(sin(sin((((1-x)*(x/1))/((x-1)*(x-x)))))*(sin((((1+1)*(1*x))/((x/x)+sin(1))))*((((1/x)/(x-x))*((x+1)/(x/x)))*(((x*x)+(x/x))+((1/1)*(1*1))))))`

cannnot be parsed by sympy (see minimum working example in code below)

Sympy hits python's maximum recursion depth
`RuntimeError: maximum recursion depth exceeded`

If I increase the recursion limit as described in [Gotchas and Pitfalls](http://docs.sympy.org/dev/gotchas.html), python crashes with a segfault
`Segmentation fault (core dumped)`

I can increase the maximum stack size for the python process using
 `ulimit -s <stack_size>`
as suggested [here](http://stackoverflow.com/questions/10035541/what-causes-a-python-segmentation-fault). This delays the segfault to larger recursion limits (e.g. 100k), but as I can't go beyond 16000 kb stack size on my system, the example above still cannot be parsed  (and it doesn't feel good, either).

The above example string seems only moderately complex to me, so I am wondering if there is any way around it?

Steps to reproduce error:

```
Python 2.7.6 (default, Jun 22 2015, 17:58:13) 
[GCC 4.8.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import sympy
>>> print sympy.__version__
1.0
>>> mystr="(sin(sin((((1-x)*(x/1))/((x-1)*(x-x)))))*(sin((((1+1)*(1*x))/((x/x)+sin(1))))*((((1/x)/(x-x))*((x+1)/(x/x)))*(((x*x)+(x/x))+((1/1)*(1*1))))))"
>>> from sympy import sin
>>> x = sympy.Symbol('x')
>>> simple = eval('sin(x)')
>>> complex = eval(mystr)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<string>", line 1, in <module>
  File "/usr/local/lib/python2.7/dist-packages/sympy/core/cache.py", line 93, in wrapper
    retval = cfunc(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/sympy/core/compatibility.py", line 899, in wrapper
    result = user_function(*args, **kwds)
[...]
  File "/usr/local/lib/python2.7/dist-packages/sympy/core/expr.py", line 112, in __neg__
    return Mul(S.NegativeOne, self)
  File "/usr/local/lib/python2.7/dist-packages/sympy/core/cache.py", line 93, in wrapper
    retval = cfunc(*args, **kwargs)
RuntimeError: maximum recursion depth exceeded
```

increasing the recursion limit:

```
>>> import sys
>>> sys.setrecursionlimit(50000)
>>> complex = eval(mystr)
Segmentation fault (core dumped)

```

**Comment by wohnjayne** [comment]
Looks like this is the same issue as  #9449

**Comment by wohnjayne** [comment]
The issue could be related to a division by zero:

The above bug happens for this short example with a cosine:
```
>>> mystr="cos((((x/x)-(x*1))/((1-1)/(1+x))))"
>>> x=eval(mystr)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<string>", line 1, in <module>
  File "/usr/local/lib/python2.7/dist-packages/sympy/core/cache.py", line 93, in wrapper
    retval = cfunc(*args, **kwargs)
[...]
  File "/usr/local/lib/python2.7/dist-packages/sympy/core/cache.py", line 93, in wrapper
    retval = cfunc(*args, **kwargs)
RuntimeError: maximum recursion depth exceeded

```
 Notice the `(1-1)` term in the denominator.

With a slightly modified term  ( `(1-2)`  instead of  `(1-1)`) the line gets evaluated properly:
```
>>> mystr="cos((((x/x)-(x*1))/((1-2)/(1+x))))"
>>> x=eval(mystr)
>>> print x
cos((-x - 1)*(-x + 1))
```

**Comment by oscarbenjamin** [comment]
This now completes without recursion error but with a zoo in the result:
```julia
In [12]: import sympy                                                                                                                          

In [13]: >>> mystr="(sin(sin((((1-x)*(x/1))/((x-1)*(x-x)))))*(sin((((1+1)*(1*x))/((x/x)+sin(1))))*((((1/x)/(x-x))*((x+1)/(x/x)))*(((x*x)+(x/x))
    ...: +((1/1)*(1*1))))))" 
    ...: >>> from sympy import sin 
    ...: >>> x = sympy.Symbol('x') 
    ...: >>> simple = eval('sin(x)') 
    ...: >>> complex = eval(mystr)                                                                                                             

In [14]: complex                                                                                                                               
Out[14]: 
            ⎛ 2      ⎞    ⎛   2⋅x    ⎞                        
zoo⋅(x + 1)⋅⎝x  + 2.0⎠⋅sin⎜──────────⎟⋅sin(sin(zoo⋅x⋅(1 - x)))
                          ⎝sin(1) + 1⎠                        
──────────────────────────────────────────────────────────────
                              x  
```

**Comment by wohnjayne** [comment]
Wow, you actually dug that one up :+1: 

I tried both examples from again on my current sympy version (1.1.1), and I am is still getting the recursion error, but I guess your comment above refers to more current (dev?) version.

**Comment by oscarbenjamin** [comment]
I'm referring to current master (1.4dev).

I find it hard to parse the expression myself. Should there be a zoo in there?

**Comment by wohnjayne** [comment]
I share your feelings about the expression :)

I currently have no way of testing it on the latest version, but since you say there is no recursion depth anymore I think we can close the ticket.

Again: thumbs up!
